### DISCUSSION TITLE
How to add Validations beforeSave hook and return error message

### DISCUSSION THREAD

**Discussion by khanakia** [filamentphp/filament#3412] [Help]
I can validate and add die() code. But how do i add a Nice Error Message?

**Comment by danharrin** [comment]
What's the use case

**Reply by khanakia** [reply]
@khanakia I have a Payment Withdrawal Form and before the user submits i want to check the user Wallet must have sufficient Balance to make a withdrawal.

**Reply by danharrin** [reply]
Okay, add a custom validation rule to the amount field?

**Reply by khanakia** [reply]
@danharrin If i add validation on the amount text change field and check if they have sufficient balance to make a withdrawal then it can be hacked easily

Here's are the steps how to hack:
1. Open the withdrawal form in multiple browsers
2. Now enter the amount to withdraw let's say they have 10 USD in their wallet so they will enter 10 USD in both forms opened in different browsers.
3. As we are validating only on the amount change field now the user can submit both forms. LOL

So filament should provide a function let say **beforeSubmit** and validate the whole form instead of validating on text change etc.

**Reply by danharrin** [reply]
This is wrong, the validation happens on submission, just before the withdrawal, not while they use the form. There is no way for the user to go overdrawn using validation.

**Reply by khanakia** [reply]
let me test it

**Reply by khanakia** [reply]
@danharrin Yes that seems to work. There is one more issue i cannot get the some fields value when `dehydrateStateUsing` is used.
As you can see in below example **beforeCreate** hooks returns Null value for the `$this->data['company_id']`
e.g.
```
Forms\Components\Hidden::make('company_id')
                    ->visibleOn('create')
                    ->dehydrateStateUsing(fn ($state) => auth()->user()->company_id),
```


```
class CreateUserPaymentRequest extends CreateRecord
{
    protected static string $resource = UserPaymentRequestResource::class;

    protected function beforeCreate(): void
    {
        var_dump($this->data['company_id']); // Returns NULL

    }
}
```

**Reply by khanakia** [reply]
@danharrin Ignore my last message, please. I was able to get the data using `$data = $this->form->getState();`
